src/main/java/io/fabric8/maven/docker/assembly/DockerFileBuilder.java

```diff
@@ -195,28 +195,26 @@ private static void buildOption(StringBuilder b, DockerFileOption option, Object
 
     private void addCopy(StringBuilder b) {
         if (assemblyUser != null) {
-            String tmpDir = createTempDir();
-            addCopyEntries(b, tmpDir);
-
-            String[] userParts = StringUtils.split(assemblyUser, ":");
-            String userArg = userParts.length > 1 ? userParts[0] + ":" + userParts[1] : userParts[0];
-            String chmod = "chown -R " + userArg + " " + tmpDir + " && cp -rp " + tmpDir + "/* / && rm -rf " + tmpDir;
-            if (userParts.length > 2) {
-                DockerFileKeyword.USER.addTo(b, "root");
-                DockerFileKeyword.RUN.addTo(b, chmod);
-                DockerFileKeyword.USER.addTo(b, userParts[2]);
-            } else {
-                DockerFileKeyword.RUN.addTo(b, chmod);
+            String[] userParts = assemblyUser.split(":");
+
+            for (CopyEntry entry : copyEntries) {
+                String dest = (basedir.equals("/") ? "" : basedir) + "/" + entry.destination;
+                if (userParts.length > 2) {
+                    DockerFileKeyword.USER.addTo(b, "root");
+                }
+                DockerFileKeyword.ADD.addTo(b, " --chown="
```

Comment:
Can't we just keep the `COPY` semnatics (instead of using `ADD` which works slightly different). I.e. I would add the extra command for chown to `addCopyEntries` and add an extra argument `String ownerAndGroup` to `addCopyEntries()` which can be either `null` or `"user:group"`. That way we can preserve the `COPY` semantics.

i.e. use `COPY --chown` instead of `ADD --chown`.

Reply:
@rohanKanojia I'd like to add this to the release. Do you think there is a chance to get this in for the next release  (i.e. change from `ADD` to `COPY`) ?

Reply:
ah, sorry. I was busy in last few weeks. Let me update this today

Reply:
Awesome, thanks !